### ISSUE TITLE
Enabling snapshot-on-shutdown feature causes panic

### ISSUE THREAD

**Issue by pstibrany** [prometheus/prometheus#9437]
**What did you do?** Enabled snapshot-on-shutdown feature via TSDB flags (`EnableMemorySnapshotOnShutdown`).

**What did you expect to see?** Not panic.

**What did you see instead? Under which circumstances?** Panic due to double registration of metrics:

```
level=info ts=2021-10-04T06:25:14.164655092Z caller=ingester_v2.go:1836 msg="successfully opened existing TSDBs"
panic: duplicate metrics collector registration attempted

goroutine 340 [running]:
github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc0004981e0, 0xc01266a540, 0x6, 0x6)
	/backend-enterprise/vendor/github.com/prometheus/client_golang/prometheus/registry.go:403 +0xb7
github.com/prometheus/prometheus/tsdb.NewExemplarMetrics(0x30219d0, 0xc0004981e0, 0x6)
	/backend-enterprise/vendor/github.com/prometheus/prometheus/tsdb/exemplar.go:95 +0x5a5
github.com/prometheus/prometheus/tsdb.(*Head).resetInMemoryState(0xc00102a5a0, 0xc024ae8640, 0x4)
	/backend-enterprise/vendor/github.com/prometheus/prometheus/tsdb/head.go:214 +0x45
github.com/prometheus/prometheus/tsdb.(*Head).Init(0xc00102a5a0, 0x17c49755a00, 0x0, 0x0)
	/backend-enterprise/vendor/github.com/prometheus/prometheus/tsdb/head.go:481 +0x217b
github.com/prometheus/prometheus/tsdb.open(0xc0008e31f0, 0xf, 0x7f257a411100, 0xc000498230, 0x30219d0, 0xc0004981e0, 0xc000512980, 0xc0008d4050, 0x1, 0xa, ...)
	/backend-enterprise/vendor/github.com/prometheus/prometheus/tsdb/db.go:760 +0x918
github.com/prometheus/prometheus/tsdb.Open(0xc0008e31f0, 0xf, 0x7f257a411100, 0xc000498230, 0x30219d0, 0xc0004981e0, 0xc000512980, 0x0, 0x203000, 0x55, ...)
	/backend-enterprise/vendor/github.com/prometheus/prometheus/tsdb/db.go:586 +0xcc
...
```
* Prometheus version:

- Prometheus main, commit https://github.com/prometheus/prometheus/commit/5a754bc043b8e8c2d57cf906386734a73761e2a2

**Comment by codesome** [comment]
This was the snapshot replay log along with it
```
ts=2021-10-04T08:07:14.719621964Z caller=log.go:168 org_id=9960 level=error msg="Failed to load chunk snapshot" err="decode loop: append exemplar: exemplar storage is disabled or max exemplars is less than or equal to 0"
```
I will fix both the panic and this failed replay where it should not hard fail on such errors and cut a v2.30.3

**Comment by codesome** [comment]
Closed by https://github.com/prometheus/prometheus/pull/9438, the fix will be merged into `main` branch after we cut v2.30.3
